# genotyphi

Assign genotypes to Salmonella Typhi genomes based on VCF files (mapped to Typhi CT18 reference genome)

## Basic Usage

```
python genotyphi.py --ref AL513382 --vcf *vcf > genotypes.txt
```

### Required options

```
--vcf   Path to one or more VCF files, generated by mapping to the Typhi CT18 reference genome (AL513382)
        Note the SNP coordinates used here for genotyping are relative to Typhi CT18 (AL513382) 
        so the input MUST be a VCF obtained via mapping to this reference sequence.

--ref   Name of the Typhi CT18 chromosome reference in your VCF file.
        This is the entry in the first column (#CHROM) of the data part of the file.
        This is necessary in case you have mapped to multiple replicons (e.g. chromosome and
        plasmid) and all the results appear in the same VCF file.
```

### Other options

```
--phred       Minimum phred quality to count a variant call vs CT18 as a true SNP (default 20)

--min_prop    Minimum proportion of reads required to call a SNP (default 0.1)
```

## Outputs

Output is to standard out, in tab delimited format.

The script works by looking for SNPs that define three nested levels of phylogenetic lineages for S. Typhi: primary clusters, clades and subclades. These are named in the form 1.2.3, which indicates primary cluster 1, clade 1.2, and subclade 1.2.3.

All genotypes implicated by the detected SNPs will be reported (comma separated if multiple are found). 

Usually subclade, clade and primary cluster SNPs will be compatible, i.e. you would expect to see 1.2.3 in the subclade column, 1.2 in the clade column and 1 in the primary clade column.

As recombination is extremely rare in S. Typhi, it is unlikely that DNA isolated from a single S. Typhi culture would have high quality SNPs that are compatible with multiple genotypes, or incompatible clade/subclade combinations... therefore if you see this in the output, it is worth investigating further whether you have contaminated sequence data or a genuine recombination. To provide very basic assistance with this, the last 3 columns indicate, for each genotype-defining SNP detected, the proportion of reads that support the genotype call. A genuine recombinant strain would have p=1 for multiple incompatible SNPs, whereas a mixed DNA sample might have p~0.5.

WARNING: Note the reference genome CT18 has the genotype 3.2.1. It is therefore possible that unexpected behaviour (e.g. wrong reference, unspecified reference name, or problems that results in overall poor quality SNP calls in the VCF) can sometimes result in calls of 3.2.1 ... so if you see lots of strains being assigned this genotype, it is worth investigating further.
